zoo_frontend/src/pages/admin/groupDiets/groupDiets.jsx

```diff
@@ -0,0 +1,237 @@
+
+const generateStateData = (subenclosures, locations) => {
+  let locationLookupIds = {};
+  locationLookupIds[0] = 'None';
+
+  locationLookupIds = locations.slice(0).reduce((acc, location) => {
+    acc[location.locationId] = location.location;
+    return acc;
+  }, locationLookupIds);
+
+  let locationLookup = {};
+  locationLookup.None = 'None';
+
+  locationLookup = locations.slice(0).sort((a, b) => (a.location > b.location ? 1 : -1)).reduce((acc, location) => {
+    acc[location.location] = location.location;
+    return acc;
+  }, locationLookup);
+
+  const subenclosuresData = subenclosures.map((sub) => ({
+    ...sub,
+    locationId: sub.locationId || 0,
+    locationName: locationLookupIds[sub.locationId || 0],
+  }));
+
+  return {
+    subenclosuresData,
+    locationLookup,
+  };
+};
+
+class GroupDiets extends Component {
+  /**
+   * Server side data retrieval
+   */
+  static async getInitialProps({ authToken }) {
+    const departmentsApi = new DepartmentAPI(authToken);
+    const subenclosuresApi = new SubenclosuresAPI(authToken);
+    try {
+      const [locRes, subenclosuresRes] = await Promise.all([
+        departmentsApi.getDepartments(),
+        subenclosuresApi.getSubenclosures(),
+      ]);
+      return {
+        subenclosures: subenclosuresRes.data,
+        locations: locRes.data,
+      };
+    } catch (err) {
+      return {
+        subenclosures: [],
+        locations: [],
+        error: true,
+        errorMessage: 'Error loading data',
+      };
+    }
+  }
+  static propTypes = {
+    // account: PropTypes.object.isRequired,
+    token: PropTypes.string,
+    classes: PropTypes.object.isRequired,
+    locations: PropTypes.arrayOf(PropTypes.object).isRequired,
+    subenclosures: PropTypes.arrayOf(PropTypes.object).isRequired,
+    error: PropTypes.bool,
+    errorMessage: PropTypes.string,
+  };
+
+  static defaultProps = {
+    token: '',
+    error: false,
+    errorMessage: '',
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      ...generateStateData(props.subenclosures, props.locations),
+    };
+    this.notificationsRef = React.createRef();
+  }
+
+  onRowAdd = (newData) => new Promise(async (resolve, reject) => {
+    const subenclosuresApi = new SubenclosuresAPI(this.props.token);
+    // Reject if a field is not filled out
+    if (!newData.subenclosure || !newData.locationName) {
+      this.notificationsRef.current.showNotification('error', 'Please fill out all of the "Species" and "Scientific Name".');
+      reject();
+      return;
+    }
+
+    // Find the selected location
+    let location = this.props.locations.find((l) => l.location === newData.locationName);
+
+    if (!location) {
+      if (newData.locationName === 'None') {
+        location = { locationId: null };
+      } else {
+        reject();
+        return;
+      }
+    }
+
+    try {
+      // Create the group diet entry
+      await subenclosuresApi.createSubenclosures({
```

Comment:
I'd prefer you take the response back from the API (the new record) and spread it into the state. Knocks off an API call